### MovieLens Case Study

The GroupLens Research Project is a research group in the Department of Computer Science and Engineering at the University of Minnesota. The data is widely used for collaborative filtering and other filtering solutions. However, we will be using this data to act as a means to demonstrate our skill in using Python to “play” with data.

### Datasets Information:

- **ratings.csv:** It contains information on ratings given by the users to a particular movie. Columns: user id, movie id, rating, timestamp

- **movie.csv:** File contains information related to the movies and its genre. Columns: movie id, movie title, release date, unknown, Action, Adventure, Animation, Children’s, Comedy, Crime, Documentary, Drama, Fantasy, Film-Noir, Horror, Musical, Mystery, Romance, Sci-Fi, Thriller, War, Western

- **user.csv:** It contains information of the users who have rated the movies. Columns: user id, age, gender, occupation, zip code

### Objective:

`To extract insights from the dataset`

### Learning Outcomes:
`Use of Pandas Functions - shape, describe, groupby, merge etc.`


#### Domain 
`Internet and Entertainment`

**Note that the case study will need you to apply the concepts of groupby and merging extensively.**

### 1. Import the necessary packages

In [1]:
import pandas as pd
import numpy as np

### 2. Read all the three datasets

In [2]:
# Reading datasets by using read_csv from pandas package
ratings = pd.read_csv("ratings.csv")
movie = pd.read_csv("movie.csv")
user = pd.read_csv("user.csv")

### 3. View the first 5 rows of all the datasets.
`Note that you will need to do it for all the three datasets seperately`

In [3]:
ratings.head(5)

user id  movie id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

In [4]:
movie.head(5)

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

In [5]:
user.head(5)

user id  age gender  occupation zip code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

### 4. Understand the shape of all the datasets.
`Note that you will need to do it for all the three datasets seperately`

In [6]:
# ratings
ratings.shape

(100000, 4)

 **Observation:** There are 100000 rows and 4 columns in the ratings dataset

In [7]:
# user
user.shape

(943, 5)

 **Observation:** There are 943 rows and 5 columns in the user dataset

In [8]:
# movie
movie.shape

(1680, 21)

 **Observation:** There are 1680 rows and 21 columns in the movie dataset

### 5. Check the data types of the columns for all the datasets.
 `Note that you will need to do it for all the three datasets seperately`

In [9]:
# ratings
# We use dataframe.dtypes to get the data types of each column
ratings.dtypes 

user id      int64
movie id     int64
rating       int64
timestamp    int64
dtype: object

 **Observation:** All columns have integer data type 

In [10]:
# user
user.dtypes

user id        int64
age            int64
gender        object
occupation    object
zip code      object
dtype: object

 **Observations:**
 1. user id and age columns are of integer data types 
 2. gender, occupation and zip code columns are of string data type

In [11]:
# movie
movie.dtypes

movie id         int64
movie title     object
release date    object
Action           int64
Adventure        int64
Animation        int64
Childrens        int64
Comedy           int64
Crime            int64
Documentary      int64
Drama            int64
Fantasy          int64
Film-Noir        int64
Horror           int64
Musical          int64
Mystery          int64
Romance          int64
Sci-Fi           int64
Thriller         int64
War              int64
Western          int64
dtype: object

**Observation:**
1. movie title and release date are of string data type
2. movie id and all genres are of interger data type

### 6. Give a statistical summary for all the datasets.
`Note that you will need to do it for all the three datasets seperately`

In [12]:
# ratings
ratings.describe()

user id       movie id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

**Observation:** Mean and median user ratings are 3.53 & 4.00 respectively

In [13]:
# user
user.describe()

user id         age
count  943.000000  943.000000
mean   472.000000   34.051962
std    272.364951   12.192740
min      1.000000    7.000000
25%    236.500000   25.000000
50%    472.000000   31.000000
75%    707.500000   43.000000
max    943.000000   73.000000

**Observation:** The average age of all the users is 34 years while the range lies between 7 to 73 years.

In [14]:
# movie
movie.describe()

movie id       Action    Adventure    Animation    Childrens  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean    841.525595     0.149405     0.080357     0.025000     0.072619   
std     485.609591     0.356593     0.271926     0.156171     0.259587   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%     421.750000     0.000000     0.000000     0.000000     0.000000   
50%     841.500000     0.000000     0.000000     0.000000     0.000000   
75%    1261.250000     0.000000     0.000000     0.000000     0.000000   
max    1682.000000     1.000000     1.000000     1.000000     1.000000   

            Comedy        Crime  Documentary        Drama      Fantasy  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.300595     0.064881     0.029762     0.431548     0.013095   
std       0.458653     0.246389     0.169980     0.495440     0.113717   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

         Film-Noir       Horror      Musical      Mystery      Romance  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.014286     0.054762     0.033333     0.036310     0.147024   
std       0.118701     0.227583     0.179559     0.187115     0.354235   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

            Sci-Fi     Thriller          War      Western  
count  1680.000000  1680.000000  1680.000000  1680.000000  
mean      0.060119     0.149405     0.042262     0.016071  
std       0.237778     0.356593     0.201246     0.125788  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000

**Observation:** The genres should be in categorical format and not in the numeric because it is of binary class

### 7. Find the number of movies per genre using the movie data

In [15]:
# Getting all the column names
movie.columns 

Index(['movie id', 'movie title', 'release date', 'Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [46]:
# This selects every 3rd column starting from the 4th column and ending at the 10th column
movie.columns[4:10:3]

Index(['movie id', 'Action', 'Childrens', 'Documentary', 'Film-Noir',
       'Mystery', 'Thriller'],
      dtype='object')

In [35]:
# Taking all the genre columns and finding the sum for every column
# movie.columns[3:] selects all the columns in the dataset that correspond to movie genre
movie[movie.columns[3:]].sum() 

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

In [17]:
# Alternatively, we can also loc function
movie.loc[:,'Action':'Western'].sum()

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

In [18]:
# Sorting the movies across genres
number = movie.loc[:,'Action':'Western'].sum()
number.sort_values(ascending = False)

Drama          725
Comedy         505
Action         251
Thriller       251
Romance        247
Adventure      135
Childrens      122
Crime          109
Sci-Fi         101
Horror          92
War             71
Mystery         61
Musical         56
Documentary     50
Animation       42
Western         27
Film-Noir       24
Fantasy         22
dtype: int64

**Observations:**
1. Drama and Comedy are the most common movie genre.
2. Clearly, there are some movies that have more than one genre.

### 8. Find the movies that have more than one genre
`Hint: use sum on the axis = 1`

In [19]:
# Checking column names
movie.columns

Index(['movie id', 'movie title', 'release date', 'Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [20]:
new_movie = movie[['movie id', 'movie title']]

In [21]:
new_movie["Number of Genres"] = movie[['movie title', 'Action',
       'Adventure', 'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary',
       'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery',
       'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']].sum(axis=1)

<ipython-input-21-34e8391b995a>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_movie["Number of Genres"] = movie[['movie title', 'Action',


In [22]:
# Filtering movies that have more than 1 genres
new_movie[new_movie['Number of Genres']>1]

movie id         movie title  Number of Genres
0            1          Toy Story                  3
1            2          GoldenEye                  3
3            4         Get Shorty                  3
4            5            Copycat                  3
6            7     Twelve Monkeys                  2
...        ...                 ...               ...
1666      1669  MURDER and murder                  3
1667      1670            Tainted                  2
1670      1673             Mirage                  2
1676      1679          B. Monkey                  2
1677      1680      Sliding Doors                  2

[849 rows x 3 columns]

**Observation:** 849 movies have more than one genre.

### 9. Find the top 25 movies according to average ratings such that each movie has number of ratings more than 100

Hint : 

1. First find the movies that have more than 100 ratings(use groupby and count). Extract the movie id in a list.
2. Find the average rating of all the movies and sort them in the descending order. 
3. Use isin(list obtained from 1) to filter out the movies which have more than 100 ratings.
4. You will have to use the .merge() function to get the movie titles.

Note: This question will need you to research about groupby and apply your findings. You can find more on groupby on https://realpython.com/pandas-groupby/.

In [23]:
# Merging ratings dataset with movie dataset
df_merge = movie.merge(ratings, on = 'movie id', how = 'inner')
# df_merge = movie.merge(ratings, on = 'movie id')

df_merge.head()

movie id movie title release date  Action  Adventure  Animation  Childrens  \
0         1  Toy Story      1-Jan-95       0          0          1          1   
1         1  Toy Story      1-Jan-95       0          0          1          1   
2         1  Toy Story      1-Jan-95       0          0          1          1   
3         1  Toy Story      1-Jan-95       0          0          1          1   
4         1  Toy Story      1-Jan-95       0          0          1          1   

   Comedy  Crime  Documentary  ...  Musical  Mystery  Romance  Sci-Fi  \
0       1      0            0  ...        0        0        0       0   
1       1      0            0  ...        0        0        0       0   
2       1      0            0  ...        0        0        0       0   
3       1      0            0  ...        0        0        0       0   
4       1      0            0  ...        0        0        0       0   

   Thriller  War  Western  user id  rating  timestamp  
0         0    0        0      308       4  887736532  
1         0    0        0      287       5  875334088  
2         0    0        0      148       4  877019411  
3         0    0        0      280       4  891700426  
4         0    0        0       66       3  883601324  

[5 rows x 24 columns]

In [24]:
# Checking the dimensions of the merged dataframe
df_merge.shape

(99990, 24)

In [25]:
# Finding the count of ratings for each movie using groupby() and count()
# reset_index() is used to shift movie title from being the dataframe’s (movie_count’s) index to 
# being just a normal column 
movie_count = df_merge.groupby(['movie title'])['rating'].count().reset_index()
movie_count.head()

movie title  rating
0  'Til There Was You        9
1               1-900        5
2      101 Dalmatians      109
3        12 Angry Men      125
4                  187      41

In [26]:
# Extracting the movie titles that have more than 100 ratings 
movie_100 = movie_count[movie_count['rating']>100]['movie title']
movie_100.head()

2            101 Dalmatians 
3              12 Angry Men 
7     2001: A Space Odyssey 
15           Absolute Power 
16               Abyss, The 
Name: movie title, dtype: object

In [27]:
# Finding average ratings for each movie and sorting them out in descending order,
# using groupby() and sort_values() on merged data frame
avg_rating = df_merge.groupby(['movie title'])['rating'].mean().sort_values(ascending=False).reset_index()
avg_rating

movie title  rating
0                  Saint of Fort Washington, The      5.0
1     Entertaining Angels: The Dorothy Day Story      5.0
2                                  Aiqing wansui      5.0
3                             Santa with Muscles      5.0
4                                    Prefontaine      5.0
...                                           ...     ...
1652         Tigrero: A Film That Was Never Made      1.0
1653                               Low Life, The      1.0
1654                              Pharaoh's Army      1.0
1655                        To Cross the Rubicon      1.0
1656                                    Power 98      1.0

[1657 rows x 2 columns]

In [28]:
# Extracting movie titles that have more than 100 ratings using movie titles in movie_100 and isin() function
# Displaying top 25 rows only
avg_rating[avg_rating['movie title'].isin(movie_100)].head(25)

movie title    rating
15                                    Close Shave, A   4.491071
16                                  Schindler's List   4.466443
17                               Wrong Trousers, The   4.466102
18                                        Casablanca   4.456790
20                         Shawshank Redemption, The   4.445230
21                                       Rear Window   4.387560
22                               Usual Suspects, The   4.385768
23                                         Star Wars   4.358491
24                                      12 Angry Men   4.344000
28                                      Citizen Kane   4.292929
30                             To Kill a Mockingbird   4.292237
31                   One Flew Over the Cuckoo's Nest   4.291667
32                         Silence of the Lambs, The   4.289744
33                                North by Northwest   4.284916
34                                    Godfather, The   4.283293
35                                    Secrets & Lies   4.265432
36                                 Good Will Hunting   4.262626
37                         Manchurian Candidate, The   4.259542
38  Dr. Strangelove or: How I Learned to Stop Worr...  4.252577
39                           Raiders of the Lost Ark   4.252381
40                                           Vertigo   4.251397
44                                           Titanic   4.245714
45                                Lawrence of Arabia   4.231214
47                               Maltese Falcon, The   4.210145
48                          Empire Strikes Back, The   4.204360

### 10. See gender distribution across different genres check for the validity of the below statements

* Men watch more drama than women
* Women watch more Sci-Fi than men
* Men watch more Romance than women

**compare the percentages**

1. There is no need to conduct statistical tests around this. Just **compare the percentages** and comment on the validity of the above statements.

[ this is not entirely accurate ]

2. you might want ot use the .sum(), .div() function here.
3. Use number of ratings to validate the numbers. For example, if out of 4000 ratings received by women, 3000 are for drama, we will assume that 75% of the women watch drama.

In [29]:
# Merging user dataset with movie and ratings(already merged : df_merge) dataset
df_merge_all = df_merge.merge(user, on = 'user id', how = 'inner')

In [30]:
# Group by gender and aggregate with sum, selecting all the genre columns 
Genre_by_gender = df_merge_all.groupby('gender').sum().loc[:,'Action':'Western'] 

In [31]:
# Add Row total of the dataframe, to get the total number of Males and Females who gave ratings
Genre_by_gender['total'] = df_merge_all['gender'].value_counts()  

In [32]:
Genre_by_gender.T

gender           F      M
Action        5442  20147
Adventure     3141  10612
Animation      995   2610
Childrens     2232   4950
Comedy        8068  21764
Crime         1794   6261
Documentary    187    571
Drama        11008  28887
Fantasy        363    989
Film-Noir      385   1348
Horror        1197   4120
Musical       1442   3512
Mystery       1314   3931
Romance       5858  13603
Sci-Fi        2629  10101
Thriller      5086  16786
War           2189   7209
Western        371   1483
total        25738  74252

In [33]:
# Divide each cell with row total and multiply by 100 to get the percentage
(Genre_by_gender.div(Genre_by_gender.total, axis= 0) * 100).T

gender                F           M
Action        21.143834   27.133276
Adventure     12.203745   14.291871
Animation      3.865879    3.515057
Childrens      8.672002    6.666487
Comedy        31.346647   29.310995
Crime          6.970239    8.432096
Documentary    0.726552    0.769003
Drama         42.769446   38.904003
Fantasy        1.410366    1.331951
Film-Noir      1.495843    1.815439
Horror         4.650711    5.548672
Musical        5.602611    4.729839
Mystery        5.105292    5.294133
Romance       22.760121   18.320045
Sci-Fi        10.214469   13.603674
Thriller      19.760665   22.606798
War            8.504934    9.708829
Western        1.441448    1.997253
total        100.000000  100.000000

**Observations:**
    
1. Of all the ratings given by women 43% were for Drama movies and for Men it was 39%. Therefore the statement is false. Women watch more Drama


2. Of all the ratings given by women 10% were for Sci-FI movies and for Men it was 14%. Therefore the statement is false. Men watch more Sci-Fi


3. Of all the ratings given by women 23% were for Drama movies and for Men it was 18%. Therefore the statement is false. Women watch more Romance

#####  Answer: All the statements were false. 